annotation/util.scm

```diff
@@ -164,21 +165,55 @@
       ((single) single)
       ((first second . rest) second))))
 
+; ----------------------------------------------------
+
+(define run-query-mtx (make-mutex))
 (define-public (run-query QUERY)
 "
   Call (cog-execute! QUERY), return results, delete the SetLink.
   This avoids a memory leak of SetLinks
 "
 	; Run the query
 	(define set-link (cog-execute! QUERY))
-	; Get the query results
-	(define results (cog-outgoing-set set-link))
-	; Delete the SetLink
-	(cog-delete set-link)
-	; Return the results.
-	results
+
+	(lock-mutex run-query-mtx)
+	(if (cog-atom? set-link)
+		; Get the query results
+		(let ((results (cog-outgoing-set set-link)))
+			; Delete the SetLink
+			(cog-delete set-link)
+			(unlock-mutex run-query-mtx)
+			; Return the results.
+			results)
+		; Try again
+		(begin
+			(unlock-mutex run-query-mtx)
+			(run-query QUERY))
+	)
+)
+
+; ----------------------------------------------------
+
+(define-public (memoize-function-call FUNC)
```

Comment:
`memoize-function-call` is defined right here @Habush

Reply:
Yeah, but for some inexplicable reason it fails saying that function is not found. I just rebuild the CI manually and it fails to compile the code.